### ISSUE TITLE
Resize consistently causes Processing error: RuntimeError: unreachable

### ISSUE THREAD

**Issue by JNavas2** [GoogleChromeLabs/squoosh#988]
**Describe the bug**
Resize consistently (100%) causes:

> Processing error: RuntimeError: unreachable

**To Reproduce**
1. Open squoosh.app in Android Chrome.
2. Open first sample image.
3. Expand lower control.
4. Turn on Resize

**Expected behavior**
Resizing of image.

**Version:**
Chrome 89.0.4389.105
Device model: Google Pixel 4a 5G (bramble)
Android version: 11 (RQ2A.210405.005)
Carrier: T-Mobile USA

**Is your issue related to the quality of image compression?**
No.

**Additional context, screenshots, screencasts**

**[YouTube video demonstrating the problem](https://youtu.be/X9dA370xevk).**

![Screenshot_20210416-081603](https://user-images.githubusercontent.com/14850792/115051411-a45df680-9e91-11eb-8313-3e0e0560a586.jpg)

**Comment by JNavas2** [comment]
YouTube failed to process original video.
**New [YouTube video demonstrating the problem](https://youtu.be/5p28kl6dCTg)**

**Comment by xfalcox** [comment]
I'm seeing this quite frequently in Android, while it works fine on Chrome/Linux, Firefox/Linux and Safari/iOS.

**Comment by xfalcox** [comment]
Disabling both "Linear RGB" and "Premultiply Alpha Channel" 'fixes' it so it should be related to

https://github.com/GoogleChromeLabs/squoosh/blob/dev/codecs/resize/src/lib.rs#L83-L96

**Comment by surma** [comment]
@xfalcox Can you share an image that crashes for you and some details about browser + phone specs?

**Comment by JNavas2** [comment]
@surma 
> Can you share an image that crashes for you and some details about browser + phone specs?

I provided that information when I opened this Issue, including a video.
Here's another video I just made: [Squoosh Processing error: RuntimeError: unreachable](https://youtu.be/dm6WGFk1o0A)
**_100% reproducible. Crashes with the provided sample image_**
Device model: Google Pixel 4a 5G (bramble)
Android version: 11 (RQ3A.210605.005)

**Comment by surma** [comment]
@JNavas2 I know you did. I was asking @xfalcox.

**Comment by xfalcox** [comment]
> @xfalcox Can you share an image that crashes for you and some details about browser + phone specs?

The image is 
![PXL_20210621_184219988](https://user-images.githubusercontent.com/1385470/123141591-ea55ae80-d42e-11eb-9ec4-46c18a3f1350.jpg)

My phone is the Pixel 3A XL on Android 12 Beta running Chrome 91.0.4472.101.

Since squoosh.app resize function is broken at the moment (`Signature Mismatch`) I'm reproducing the problem on my app where I re-use the `resize` codec directly.

**Comment by nohe427** [comment]
**Some notes and observations I made triaging this**

I tried reproducing this on my pixel 5 today and if I wait for the image to process before I uncheck "Linear RGB" and "Premultiply Alpha Channel", I do not see this error. I can continue clicking on If I quickly click resize -> and then uncheck either linear rgb or premultiply alpha channel, I see this Signature Mismatch error.


* When I inspect with Chrome Dev Tools over ADB, I do not see this error at all. Everything works as expected. When I refresh the page with devtools closed, it is all broken again.
* When digging in, and quickly doing the uncheck either linear rgb or premultiply alpha channel to generate that error message, I can see that things start acting a little odd around [this line in the resize function](https://github.com/GoogleChromeLabs/squoosh/blob/324b04d398545864f0de9bac316404b1b6d6b8f6/codecs/resize/src/lib.rs#L104).
* Considering that each of the devices tested in this bug have different amounts of RAM assigned to them, (Pixel 4a 5g - 6gb, Pixel 5 - 8 gb, Pixel 3a XL - 4 gb) I feel like this is some sort of memory issue, especially since we are all experiencing this issue at different parts of the process.
* This makes me wonder if there could potentially be a problem with filling the vector with values. When testing on multiple different ways of filling the vector, I noticed that I would continually get some garbage output after resizing or filling (with a for loop) the vector and this is making me feel like this is potentially a buffer overflow issue based on the output that I am getting when I test this myself.

**Comment by surma** [comment]
Thanks for digging into this, @nohe427! 

> When I inspect with Chrome Dev Tools over ADB, I do not see this error at all. Everything works as expected. When I refresh the page with devtools closed, it is all broken again.

This is very curious. I might be getting ahead of ourselves, but it could be a bug in Chrome/Wasm! When DevTools is open, V8 disables the Wasm’s optimizing compiler so that you can step through your code. With DevTools closed, the optimizing compiler is active and will generated optimized (i.e. different) machine code).

> When digging in, and quickly doing the uncheck either linear rgb or premultiply alpha channel to generate that error message, I can see that things start acting a little odd around this line in the resize function.
> This makes me wonder if there could potentially be a problem with filling the vector with values. When testing on multiple different ways of filling the vector, I noticed that I would continually get some garbage output after resizing or filling (with a for loop) the vector and this is making me feel like this is potentially a buffer overflow issue based on the output that I am getting when I test this myself.

Can you elaborate a bit more on this? Technically Rust should not allow us to have buffer overflows, but you never know. What exactly are you seeing? What exactly did you try?

**Comment by nohe427** [comment]
@surma

> This is very curious. I might be getting ahead of ourselves, but it could be a bug in Chrome/Wasm! When DevTools is open, V8 disables the Wasm’s optimizing compiler so that you can step through your code. With DevTools closed, the optimizing compiler is active and will generated optimized (i.e. different) machine code).

Yeah, when I discovered that opening devtools seemed to fix everything I thought I must've been doing something wrong. Good to know that tidbit about WASM optimizing compiler being disabled.

> Can you elaborate a bit more on this? Technically Rust should not allow us to have buffer overflows, but you never know. What exactly are you seeing? What exactly did you try?

**How I tested**
I didn't do any sort of elegant debugging when testing this, but rather I just used `web_sys::console::log_1` from the `web-sys` crate. This allowed me to send messages straight to the javascript console while I ran the resize command.

**The order in which I tested**
* When I was testing without any debug statements, I found that if I wait for the image to load (the little processing symbol in the bottom right corner to stop spinning) and change resize parameters one at a time, there was no issues.
* When I would not wait for the image to load, and toggle the changes quickly, I would then see a message that the method signature did not match. I used javascripts `console.log` to validate the parameters that were being passed on each iteration to ensure that they matched and they did.
* I decided to start throwing logging statements into rust using the aforementioned `log_1` method and noticed that when initializing a vector with zero values (`    preprocessed_input_image.resize(input_image.len(), 0.0f32);`), that the next log method would change to `☐☐☐`. [link](https://github.com/nohe427/squoosh/blob/4794e0bb69117f7bdde58c50c565645eb40901f3/codecs/resize/src/lib.rs#L120)
  * Its been a little while since I worked in low level languages, but this looked oddly like a buffer overflow to me. It seems that whether I use the `resize` method or a `for-loop` to init the values in the vector, anything afterwards seems to be garbaled, no matter where in the resize function I put this.
  * When I see this, it makes me feel like the specified amount of space is overflowed when filling the vector

**Comment by nohe427** [comment]
@surma - Is there a way to individually turn on and off the WASM optimizations within DevTools in order to chase this down? Is there a larger amount of optimizations that occur with WASM like [there are with C](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html)?

**Comment by surma** [comment]
This is some great sleuthing work, @nohe427! Thank you. 

> I decided to start throwing logging statements into rust using the aforementioned log_1 method and noticed that when initializing a vector with zero values ( preprocessed_input_image.resize(input_image.len(), 0.0f32);), that the next log method would change to ☐☐☐.

This one is super interesting. Does the log also go weird when DevTools is open? 

If so, I increasingly suspect a V8 bug here rather than a Rust bug, and we should probably try and build a simple reproduction case that we can give to the V8 folks. What I think would be useful next steps here is to take out the .wasm and .js file _with_ your changes and run them via a d8 script. `d8` (installable via [jsvu](https://github.com/GoogleChromeLabs/jsvu)) allows you to run a script with and without optimizers enabled. [We already have one d8 script](https://github.com/GoogleChromeLabs/squoosh/blob/dev/codecs/rotate/benchmark.js) committed, so you can use that for inspiration.

```
# 1. Install `v8` via jsvu.
# 2. Run a script without optimizer
v8 --liftoff --no-wasm-tier-up --no-opt ./myfile.js
# 3. Run a script with only the optimized output
v8 --no-liftoff --no-wasm-tier-up ./myfile.js
```

WDYT? Does this make sense?

**Comment by nohe427** [comment]
@surma 

> This one is super interesting. Does the log also go weird when DevTools is open?

Nope. The log only goes haywire if DevTools is closed. Once DevTools is open, I cannot get the crazy logging.
 
> If so, I increasingly suspect a V8 bug here rather than a Rust bug, and we should probably try and build a simple reproduction case that we can give to the V8 folks. What I think would be useful next steps here is to take out the .wasm and .js file _with_ your changes and run them via a d8 script. `d8` (installable via [jsvu](https://github.com/GoogleChromeLabs/jsvu)) allows you to run a script with and without optimizers enabled. [We already have one d8 script](https://github.com/GoogleChromeLabs/squoosh/blob/dev/codecs/rotate/benchmark.js) committed, so you can use that for inspiration.
> 
> ```
> # 1. Install `v8` via jsvu.
> # 2. Run a script without optimizer
> v8 --liftoff --no-wasm-tier-up --no-opt ./myfile.js
> # 3. Run a script with only the optimized output
> v8 --no-liftoff --no-wasm-tier-up ./myfile.js
> ```
> 
> WDYT? Does this make sense?

It sort of makes sense, but I think I need to interrupt processing to reproduce it effectively. Let me try to explain this case a little more:

The problem that I am having is that I think the `AbortController` that exists within this project is at least partially at fault as well. If I open Squoosh and wait for the little processing symbol to stop and show save instead, I cannot repro the issue anymore. So: 
* Open Squoosh and select an image.
* Wait for processing to finish.
* Click Resize
* Wait for processing to finish.
* Uncheck one of the options mentioned in a previous comment
* Wait for processing to finish.
* Uncheck the other one
* Wait for processing to finish.

I cannot get the weird error message at this point. I am wondering if there is perhaps some sort of cleanup that comes after the `AbortController::Signal` is received that may cause the memory to get all wacky when trying to reinvoke the resize command.

The other challenge is that this is only happening on Android. Am I able to run these scripts in an Android context?

**Comment by JNavas2** [comment]
@nohe427 
I'm not seeing that work-around behavior here.

1. Launch Squoosh.
2. Open 1st (leftmost) sample image.
3. Wait for processing to stop.
4. Expand lower control.
5. Touch Resize.

_Immediate_ error (_before_ any processing).

**Comment by nohe427** [comment]
Hi @JNavas2 

Yeah, I noticed that from your YouTube video. The challenge I am having is that I only have access to Pixel 5 devices right now so cannot validate the behavior that you are seeing on my device unfortunately. Are you also unable to reproduce the issue with Chrome DevTools open on your device?

**Comment by JNavas2** [comment]
@nohe427 
What exactly would you like me to try? 
Do you mean [Remote debug Android devices](https://developer.chrome.com/docs/devtools/remote-debugging/)?

**Comment by nohe427** [comment]
@JNavas2 - yup, exactly. Does it also crash for you when doing remote debugging? Surma indicated that web assembly optimization gets disabled when devtools is open and it would be a good indicator that we're on the right path here. I'd ask that you open squoosh first, start inspecting, then go through the normal flow to generate a crash. Does that sound okay?

**Comment by JNavas2** [comment]
@nohe427 
I'll see if I can get to that tomorrow.
In the meantime, I disabled all available WebAssembly flags (as in the attached images), and that did not help.

![webassembly_flags](https://user-images.githubusercontent.com/14850792/123550665-7b74a000-d723-11eb-8a93-f39952fef028.png)

**Comment by JNavas2** [comment]
@nohe427
I confirm that with Remote Debugging on my PC, the Squoosh error did **not** occur.

**Comment by nohe427** [comment]
@JNavas2 - Thank you!

**Comment by nohe427** [comment]
@surma - I am trying to follow along with the JSVU script you mentioned I should write, but what I feel like I want to do is import the `squoosh_resize.js` file and run that with my inputs. The challenge I am having is that the TextDecoder is not defined (I believe because its a V8 runtime rather than browser) and I am curious if this is the right approach to take with writing the jsvu script. Any tips would be appreciated.

**Comment by surma** [comment]
Ah, that’s a good point @nohe427. Didn’t think about that.

I have actually learned since that Node exposes the same flags! So you can just do `node --no-liftoff` etc. Maybe that’s a way forward?

**Comment by nohe427** [comment]
@JNavas2 - Are you still able to repro? I am not able to on my Pixel 5

**Comment by JNavas2** [comment]
> @JNavas2 - Are you still able to repro? I am not able to on my Pixel 5

I checked, and I still get the error on my Pixel 4a 5G.
I also get the error in Edge.

**Comment by JNavas2** [comment]
Issue persists in Edge Android 93.0.961.38

![Screenshot_20210908-080347 (Small)](https://user-images.githubusercontent.com/14850792/132537040-ef102dc2-65a4-435d-86a4-4198980a4ca8.jpg)

**Comment by JNavas2** [comment]
**Issue persists.**
• Google Pixel 3a (sargo)
• Android 11 (RQ3A.210905.001)
• Chrome version 94.0.4606.71

**Comment by JNavas2** [comment]
The issue does **not** exist in Edge version 46.06.4.5161 (not Chromium based)
but does exist in Chromium based versions of Edge (as previously reported).
**Thus the issue is apparently caused by something in Chromium.**

**Comment by zdnk** [comment]
Any solutions yet?

**Comment by JNavas2** [comment]
New tests:
- does **still** exist in Google Chrome for Android 102.0.5005.78
- does **not** exist in Microsoft Edge for Android (Chromium based) 101.0.1210.53

So:
- It's a Chrome problem.
- Yet another reason to use Edge instead of Chrome.

**Comment by jhirvioja** [comment]
![r_resize_error](https://github.com/GoogleChromeLabs/squoosh/assets/98736266/df46bb2a-0a75-4a9d-9eaf-4fe69bbe5677)

Lot of errors when resizing with Chrome version 115.0.5790.160 on iOS 16.6. Same error with Safari.

**Comment by jasonwilliams** [comment]
I still get this on iPhone iOS 26.2, I would love to see this tackled and fixed as it prevents me sqooshing on mobile.

I don't think it's a Chrome problem as I have this on IOS which is WebKit.

Not sure if we narrowed down what the issue is or if you need reproduction steps (the same repro steps as above).
